src/scroll.js

```diff
@@ -0,0 +1,41 @@
+const AlpineScrollMagicMethod = {
+    start() {
+
+        Alpine.addMagicProperty('scroll', function ($el) {
+            return function (target) {
+                var isSmoothScrollSupported = 'scrollBehavior' in document.documentElement.style;
```

Comment:
I wonder where we should have a second argument to the function that determines whether or not scroll smooth should be used too?

Reply:
Should we mimic the default behavior with an optional third param? 

```js
$scroll(x = 0, y = 0, { behavior: 'smooth' })
```
But each being optional and support for `$scroll({ top:0, left:0, behavior: 'smooth' })`

Feels like this should be a wrapper for the native method, but provides a similar experience across all browsers

Reply:
Yeah, it could be an option. Do you think people will need "snap scrolling"?

Reply:
There will probably be one person who will need it only because we don't support it :)

Reply:
okay, I'm not sure about using a complicate syntax though. 
I would go with $scroll(whatever) doing the smooth scrolling and $scroll(whatever, false) doing snap scrolling.
If we just mimic the native syntax and we ask them to check the MDN documentation, people may as well just type `window.scrollTo({top:y, left:x, behaviour: 'smooth'})` since the DX is pretty much the same (ok, you type 8 fewer characters :D ).
I suspect part of the Alpine audience doesn't have a lot of javascript expertise and they just would like an helper with a nice syntax that just work.

Reply:
Could we support both? So it will first check whether it's an element, and if not fallback.

It might be nice to support something like `$scroll(element, {offset: '100px'})` since often people don't want it to be flush at the edge of the element too. 

We could probably support all of these cases in one go with it still feeling intuitive. They don't all have to be documented either.

Reply:
The second parameter being an object is a JS thing (I'm a backend developer, you know), I'd rather it being $scroll(target, offset, isSmooth) but I'm open to discuss it.
The only issue is that scrollTo doesn't support an offset as far as i know. In that case we would need to manually implement the scrolling with could be an headache. Unless you have something in mind already.

I think I'll be fighting over the other point, sorry. :) Is there a use case where you need to pass the native object syntax? We need to be careful to not over-engineer a solution and aim for a MVP release first. It's really hard to remove a feature once we add that in because there will be a 1% of the user using it. 
My personal approach is that if there is already a clear why to implement a feature, we shouldn't duplicate the syntax to achieve the same result. Docs will educate them. If the developer is skilled and they need something really sophisticated, they can easily implement it with a touch of vanilla JS, if they are not skilled they'll get confused by the native syntax and they'll just raise a lot of questions. It's my personal opinion though.

Reply:
I'd rather we talk it out a bit than just move forward with what's comfortable :) I'm fine either way though on a lot of this. 

On object parameters, I like to avoid `$scroll(target, offset, isSmooth)` because you then need `$scroll(target, offset, false)` and you have to remember what `false` is 6 months later. It also lets you add in more options later on without having to accept more params.

For offset, yeah, we would just need to do the math. This could skipped for now though and added later (especially if we use the object parameter approach).

On native support, the idea would be that we've included the polyfill. Ideally we are helping people who don't even know what the word polyfill is (the WP jQuery copy/paste crowd, for example, which I was when I started too so maybe I can relate). I don't think we need it though, at least not now, and it could be added later easily enough as well.

But if it's easy enough to fallback to it, then might as well.
```js
function() {
  // check if the params meet any of our special conditions

  // fall back to the default with polyfill if not
}
```

Reply:
mmh, what would you do if people writes `$scroll(target, { behavior: 'easy-in' })`? Should we throw an error?
I hardly had to change the signature for an API from having 2 parameters to having 20 parameters in 12 years.
Those massive functions usually have a bad design from the very beginning but I get what you mean.
If the majority of people prefer the object syntax we can go with that (but I think it's less nice to see in your html).
👍

Reply:
I think the primary parameters should not be objects, and only overrides from our defaults should be in there. So `$scroll(target)` and `$scroll(target, { behavior: 'easy-in' })` would be expected, and in the future `$scroll(target, { offset: 100, behavior: 'easy-in', anotherOption: true })` if we want.


So we could end up with something like this:
```js
$scroll(MIXED target, OBJECT options)
```
Where the target can be an element, a selector, or a numerical value, and options is an object with any overrides.

However, the user passes in an object as the first parameter, we pass it straight to `window.scroll` ([url](https://developer.mozilla.org/en-US/docs/Web/API/Window/scroll)) which adds the benefit of having a polyfill for Safari/IE. I don't think this part needs to be documented though. I did this with `$fetch()` actually, except it's documented.

Reply:
Yeah but easyin is not a valid value. You can only use smooth or auto. 

Btw, I think I'm always going to use win scroll anyway so the polyfill will apply regardless. It's a bit weird than you can pass $scroll({top: 10, behavior: 'scroll'}, {bahavior: auto)}) though. Which one would win? Maybe i can send a first version and you can tweak it with the object support to make it work as you think we need.

Reply:
> It's a bit weird than you can pass $scroll({top: 10, behavior: 'scroll'}, {bahavior: auto)}) though

In that case, the second parameter would be ignored. So you would have a switch statement that checks the type of the first parameter and go from there. if (param[0] is object) "handle that situation", if (element) "do another thing", etc

> Yeah but easyin is not a valid value. You can only use smooth or auto.

I thought you were implying we'd have some sort of custom easing function `ease-in` (which could be a future feature).  Does it error in `window.scroll()`? Maybe we would just ignore unknown options, or console log a warning.

Reply:
No, I was assuming that people would try.
I pushed an update but I made it in a way that the second parameters would override the first one because it's less messy code wise. I still need to update the documentation. Gtg out now, let's catch up later

Reply:
Looks great to me! I'll try it out tomorrow though. 

Do you think there's a need to support string numbers? `$scroll('300')` which could come from a variable.

Reply:
good shout. Added